src/compiler/scala/tools/nsc/backend/jvm/opt/BytecodeUtils.scala

```diff
@@ -311,13 +341,298 @@ object BytecodeUtils {
+    "scala/runtime/VolatileIntRef(I)V",
+    "scala/runtime/VolatileLongRef(J)V",
+    "scala/runtime/VolatileFloatRef(F)V",
+    "scala/runtime/VolatileDoubleRef(D)V")
+
+  def isRuntimeRefConstructor(insn: MethodInsnNode): Boolean = {
+    insn.name == INSTANCE_CONSTRUCTOR_NAME && runtimeRefConstructors(insn.owner + insn.desc)
+  }
+
+  private val tupleConstructors = Set.empty[String] ++ {
+    (1 to 22).map(n => "scala/Tuple" + n + "(" + ("Ljava/lang/Object;" * n) + ")V")
+  } ++ {
+    Iterator("I", "J", "D").map(t => "scala/Tuple1$mc" + t + "$sp(" + t + ")V")
+  } ++ {
+    def tuple2Specs = Iterator("I", "J", "D", "C", "Z")
+    for (a <- tuple2Specs; b <- tuple2Specs) yield "scala/Tuple2$mc" + a + b + "$sp(" + a + b + ")V"
+  }
+
+  def isTupleConstructor(insn: MethodInsnNode): Boolean = {
+    insn.name == INSTANCE_CONSTRUCTOR_NAME && tupleConstructors(insn.owner + insn.desc)
+  }
+
+  // unused objects created by these constructors are eliminated by pushPop
+  private val sideEffectFreeConstructors = primitiveBoxConstructors ++
+    runtimeRefConstructors ++
+    tupleConstructors ++ Set(
+    "java/lang/Object()V",
+    "java/lang/String()V",
+    "java/lang/String(Ljava/lang/String;)V",
+    "java/lang/String([C)V"
+  )
+
+  def isSideEffectFreeConstructorCall(insn: MethodInsnNode): Boolean = {
+    insn.name == INSTANCE_CONSTRUCTOR_NAME && sideEffectFreeConstructors(insn.owner + insn.desc)
+  }
+
+  private val classesForSideEffectFreeConstructors = sideEffectFreeConstructors.map(s => s.substring(0, s.indexOf('(')))
+
+  // we only eliminate `NEW C` if the class C has a constructor that we consider side-effect free.
+  // removing a `NEW` eliminates a potential NoClassDefFoundError, so we only do it for core classes.
+  def isNewForSideEffectFreeConstructor(insn: AbstractInsnNode) = {
+    insn.getOpcode == NEW && {
+      val ti = insn.asInstanceOf[TypeInsnNode]
+      classesForSideEffectFreeConstructors(ti.desc)
     }
   }
 
+  def isBoxedUnit(insn: AbstractInsnNode) = {
+    insn.getOpcode == GETSTATIC && {
+      val fi = insn.asInstanceOf[FieldInsnNode]
+      fi.owner == "scala/runtime/BoxedUnit" && fi.name == "UNIT" && fi.desc == "Lscala/runtime/BoxedUnit;"
+    }
+  }
+
+  private def buildFunctionTypes(base: String): Set[InternalName] = {
+    def primitives = Iterator("B", "S", "I", "J", "C", "F", "D", "Z", "V")
+    def ijfd = Iterator("I", "J", "F", "D")
+    def ijfdzv = Iterator("I", "J", "F", "D", "Z", "V")
+    def ijd = Iterator("I", "J", "D")
+    Set.empty[String] ++ {
+      (0 to 22).map(base + _)
+    } ++ {
+      primitives.map(base + "0$mc" + _ + "$sp") // Function0
+    } ++ {
+      // return type specializations appear first in the name string (alphabetical sorting)
+      for (r <- ijfdzv; a <- ijfd) yield base + "1$mc" + r + a + "$sp" // Function1
+    } ++ {
+      for (r <- ijfdzv; a <- ijd; b <- ijd) yield base + "2$mc" + r + a + b + "$sp" // Function2
+    }
+  }
```

Comment:
The string refs to symbols in this file, and to the specialized ones here in particular, are the main thing I think needs a cleanup before merging.